sdk/python/kfp/compiler/compiler.py

```diff
@@ -646,6 +646,10 @@ def _compile(self, pipeline_func):
     p.ops = sanitized_ops
 
     workflow = self._create_pipeline_workflow(args_list_with_defaults, p, p.conf.op_transformers)
+
+    import json
+    workflow.setdefault('metadata', {}).setdefault('annotations', {})['kubeflow.org/pipelines/pipeline_spec'] = json.dumps(pipeline_meta.to_dict())
```

Comment:
Isn't dict.setdefault returns the default {} when there is no such key?

Reply:
And, you could use pipeline_meta.serialize().

Reply:
> Isn't dict.setdefault returns the default {} when there is no such key?

Yes. And it also adds it to the dictionary. 
The behavior is like this:
```python
def setdefault(dict, key, default):
    if key not in dict:
        dict[key] = default
    return dict[key]
```

Reply:
> And, you could use pipeline_meta.serialize().

I can do that.

Reply:
If setdefault returns the value instead of the dict, why would this chained setdefault call work?

Reply:
>If setdefault returns the value instead of the dict, why would this chained setdefault call work?

`workflow.setdefault('metadata', {})` returns `{}`, but it also adds it to the `workflow` dict. Dicts are pass-by-reference, so when you add something to that `{}`, the workflow['metadata'] also gets it.

Reply:
c++ vs. python. My thinking was in C++. Got it.